nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,344 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_core/controller_exceptions.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "nav2_graceful_motion_controller/graceful_motion_controller.hpp"
+#include "nav2_costmap_2d/costmap_filters/filter_values.hpp"
+
+namespace nav2_graceful_motion_controller
+{
+
+void GracefulMotionController::configure(
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  std::string name, const std::shared_ptr<tf2_ros::Buffer> tf,
+  const std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros)
+{
+  auto node = parent.lock();
+  parent_node_ = parent;
+  if (!node) {
+    throw nav2_core::ControllerException("Unable to lock node!");
+  }
+
+  costmap_ros_ = costmap_ros;
+  tf_buffer_ = tf;
+  plugin_name_ = name;
+  logger_ = node->get_logger();
+
+  // Handles storage and dynamic configuration of parameters.
+  // Returns pointer to data current param settings.
+  param_handler_ = std::make_unique<ParameterHandler>(
+    node, plugin_name_, logger_,
+    costmap_ros_->getCostmap()->getSizeInMetersX());
+  params_ = param_handler_->getParams();
+
+  // Handles global path transformations
+  path_handler_ = std::make_unique<PathHandler>(
+    tf2::durationFromSec(params_->transform_tolerance), tf_buffer_, costmap_ros_);
+
+  // Handles the control law to generate the velocity commands
+  control_law_ = std::make_unique<SmoothControlLaw>(
+    params_->k_phi, params_->k_delta, params_->beta, params_->lambda, params_->slowdown_radius,
+    params_->v_linear_min, params_->v_linear_max, params_->v_angular_max);
+
+  // Initialize footprint collision checker
+  collision_checker_ = std::make_unique<nav2_costmap_2d::
+      FootprintCollisionChecker<nav2_costmap_2d::Costmap2D *>>(costmap_ros_->getCostmap());
+
+  // Publishers
+  transformed_plan_pub_ = node->create_publisher<nav_msgs::msg::Path>("transformed_global_plan", 1);
+  local_plan_pub_ = node->create_publisher<nav_msgs::msg::Path>("local_plan", 1);
+  motion_target_pub_ = node->create_publisher<geometry_msgs::msg::PointStamped>("motion_target", 1);
+  slowdown_pub_ = node->create_publisher<visualization_msgs::msg::Marker>("slowdown", 1);
+
+  RCLCPP_INFO(logger_, "Configured Graceful Motion Controller: %s", plugin_name_.c_str());
+}
+
+void GracefulMotionController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_.reset();
```

Comment:
Also collision checker, control law, path handler, and param handler's pointer resets